### PR TITLE
allow set_multithreading_enabled to act as function and context manager 

### LINKED ISSUE #104985: torch.autograd.set_multithreading_enabled doesn't work as a function
### 🐛 Describe the bug

The documentation for `torch.autograd.set_multithreading_enabled` says that "It can be used as a context-manager or as a function." but it doesn't work unless you store the result as it will act as a context manager and undo the value once the returned value gets destroyed.

We should either document that fact of fix `set_multithreading_enabled` behavior.

It's rather complicated to write a test for such trivial situation as there's no getter for that autograd property.

### Versions

master

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7

### PR DESCRIPTION
**Pull request by poseljacob**
Fixes #104985 

Implemented `set_multithreading_enabled` C++ function to directly alter state rather than using `MultithreadingEnabled` class, which was automatically resetting the state when the object was destroyed. This behavior more closely aligns with set_grad_enabled which does work as expected. This allows us to change python class `set_multithreading_enabled` to act as both a function and context manager. 

I also added a getter: `torch._C.is_multithreading_enabled`

### CODE REVIEW

File: torch/csrc/autograd/init.cpp [hunk 1]
```diff
@@ -754,6 +754,33 @@ static PyObject* is_fwd_grad_enabled(PyObject* _unused, PyObject* arg) {
...
+        r, args, kwargs, torch_C_module, "torch._C", "_set_multithreading_enabled");
+  }
+  auto multithreading_enabled = r.toBool(0);
+  c10::AutogradState::get_tls_state().set_multithreading_enabled(multithreading_enabled);
+  Py_RETURN_NONE;
+  END_HANDLE_TH_ERRORS
+}
+
+static PyObject* is_multithreading_enabled(PyObject* self, PyObject* args) {
+    if (c10::AutogradState::get_tls_state().get_multithreading_enabled()) {
```
Comment (albanD) [inline torch/csrc/autograd/init.cpp:777]: The HANDLE_TH_ERRORS macros are missing for this function.

**Review by albanD — COMMENTED** [review]
Thanks for sending this PR!
Added some small comments to tweak the public API but all good otherwise.

File: torch/csrc/autograd/init.cpp [hunk 2]
```diff
@@ -1004,6 +1031,11 @@ static PyMethodDef methods[] = { // NOLINT
      is_anomaly_check_nan_enabled,
      METH_NOARGS,
      nullptr},
+    {"is_multithreading_enabled", is_multithreading_enabled, METH_NOARGS, nullptr},
```
[the remark this answers is not in the document]
Comment (poseljacob) [reply torch/csrc/autograd/init.cpp:1034]: Thanks for the feedback. In which file should documentation be added via `_add_docstr`? Perhaps under `is_multithreading_enabled = torch._C._is_multithreading_enabled` in `autograd/__init__.py`?

[the remark this answers is not in the document]
Comment (albanD) [reply torch/csrc/autograd/init.cpp:1034]: Yes in the autograd init python file is good!

**Comment by albanD** [comment]
The CI failures will go away once you make the suggested change to have nicer public API.

File: torch/_C/__init__.pyi.in [hunk 3]
```diff
@@ -1159,6 +1159,8 @@ def _increment_version(tensor: Tensor) -> None: ...
 def set_anomaly_enabled(enabled: _bool, check_nan: _bool = True) -> None: ...
 def is_anomaly_enabled() -> _bool: ...
 def is_anomaly_check_nan_enabled() -> _bool: ...
+def is_multithreading_enabled() -> _bool: ...
```
Comment (albanD) [inline torch/_C/__init__.pyi.in:1162]: nit: this needs updating with the _

File: test/test_autograd.py [hunk 4]
```diff
@@ -10777,6 +10777,27 @@ def fn2(x1, x2):
         torch.autograd.gradcheck(fn2, [inp_r, inp_c], check_forward_ad=True)
         torch.autograd.gradcheck(fn2, [inp_c, inp_r], check_forward_ad=True)
 
+    def test_set_multithreading_enabled_as_context_manager_and_function(self):
+        # Test as a context manager
+        with torch.autograd.set_multithreading_enabled(False):
+            self.assertFalse(torch._C._is_multithreading_enabled())
```
Comment (albanD) [inline test/test_autograd.py:10783]: you can use the public version in the tests! `torch.autograd.is_multithreading_enabled()`

**Review by albanD — COMMENTED** [review]
Thanks for the quick update. Only two small things.

File: test/test_autograd.py (hunk 4 above)
Comment (poseljacob) [reply test/test_autograd.py]: Cool thank you, appreciate the quick review!

**Review by albanD — APPROVED** [review]
SGTM!

**Comment by poseljacob** [comment]
@albanD Thank you! If you'd like I can also work on similar fixes to classes such as `_force_original_view_tracking` which the documentation states work as both a function and context manager but do not.

**Comment by albanD** [comment]
@pytorchbot merge

>  If you'd like I can also work on similar fixes to classes such as _force_original_view_tracking which the documentation states work as both a function and context manager but do not.

Yes that would be useful to give it the same treatment to make sure it works well.
Thanks for helping with these!

**Comment by albanD** [comment]
Ho the lint issue is real, could you fix that please? You should be able to `pip install lintrunner` locally, then run `lintrunner init` and `lintrunner -a -m main`.

**Comment by poseljacob** [comment]
Hey I went ahead and did that, however, I did receive some errors locally 
```
  Error (MYPY) [attr-defined]
    Module has no attribute "_is_multithreading_enabled"; maybe
    "is_multithreading_enabled", "_set_multithreading_enabled", or
    "_MultithreadingEnabled"?

```
Kinda strange, what do you think?

**Review by albanD — APPROVED** [review]
Thanks!

**Pull request closed without merging**